psl-clj/src/psl_clj/core.clj

```diff
@@ -0,0 +1,347 @@
+     {:pre [(not-any? nil? [datastore model parts-to-read part-to-write preds-to-close])]}
+     (let [parts-to-read (into-array Partition parts-to-read)
+           preds-to-close (HashSet.
+                           (for [pname preds-to-close]
+                             (.getPredicate model pname)))]
+       (.getDatabase datastore part-to-write preds-to-close parts-to-read)))
+  
+  ;; No predicates to close
+  ([datastore model parts-to-read part-to-write]
+     {:pre [(not-any? nil? [datastore model parts-to-read part-to-write])]}
+     (let [parts-to-read (into-array Partition parts-to-read)]
+       (.getDatabase datastore part-to-write parts-to-read)))
+  
+  ;; No predicates to close or write partition
+  ([datastore model parts-to-read]
+     {:pre [(not-any? nil? [datastore model parts-to-read])]}
+     (let [parts-to-read (into-array Partition parts-to-read)]
+       (.getDatabase datastore (first parts-to-read) parts-to-read))))
+                                                                     
+(defn p                                                 
+  "Get a predicate object from the PSL model"
+  [model pred-name]
+  (.getPredicate model pred-name))
+
+;;; =============== Functions for handling PSL partitions ====================
+
+(defn partition-copy-atoms
+  "Copy atoms of given predicates from one partition to another"
+  [datastore model part-from part-to preds]
+  (let [dbr (open-db datastore model [part-from])
+        dbw (open-db datastore model [part-to] part-to)]
+    (try
+      (doseq [pnam preds]
+        (let [atoms (Queries/getAllAtoms dbr (p model pnam))]
+          (doseq [atom atoms]
+            (.commit dbw atom))))
+      nil
+      (finally (close-db dbr)
+               (close-db dbw)))))
+                                       
+(defn partition-delete                 
+  "Delete the contents of the partition in the datastore."
+  [datastore partition] (.deletePartition datastore partition))
+                                                              
+(defn partitions-delete                                       
+  "Delete the contents of the partitions in the datastore."
+  
+  ;; Selected partitions
+  ([datastore partitions]
+   (doseq [partition partitions]
+     (partition-delete datastore partition))
+   (count partitions))
+  
+  ;; All partitions
+  ([datastore]
+   (let [partitions (.listPartitions datastore)]
+     (doseq [partition partitions]
+       (partition-delete datastore partition))
+     (count partitions))))
+                    
+(defn partition-id-new
+  "Return an integer that is (probably!) unique within this session."
+  []
+  (u/uuid-int))
+
+(defn partition-new
+  "Return a unique partition within this session."
+  []
+  (let [id (partition-id-new)
+        part-name (str id)]
+    (Partition. id part-name)))
+
+;;; =============== Functions for handling data in PSL ====================
+
+(defn pred-append                                   
```

Comment:
I would also pass through some of the Inserter file loading methods.

Reply:
Agreed.  These data handling functions are geared towards making Incanter (like pandas or R, but in JVM) seamlessly interoperable with PSL.  So, to load from files you'd use the built-in capabilities of Incanter.  But I'm guessing there are advantages of Inserter, such as bulk load?